File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
// bandage = [시전 시간(t), 초당 회복량(x), 추가 회복량(y)]
// health = 최대 체력
// attacks = [공격 시간, 피해량]
function solution(bandage, health, attacks) {
    let answer = health;
    const t = bandage[0];
    const x = bandage[1];
    const y = bandage[2];
    let time = 0;
    
    for (const attack of attacks) {
        let heal = 0;
        const healTime = attack[0] - time;
        
        const additionalHeals = Math.floor(healTime / t);
        heal += (x * healTime) + (y * additionalHeals);
        
        if (answer + heal >= health) {
            answer = health;
        } else {
            answer += heal;
        }
        
        answer -= attack[1];
        if (answer <= 0) return -1;
        time = attack[0] + 1;
    }
    
    return answer;
}

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
# bandage = [시전 시간, 초당 회복량, 추가 회복량]
# health = 최대 체력
# attacks = [공격 시간, 피해량], 공격 시간을 기준으로 오름차순 정렬된 상태

def solution(bandage, health, attacks):
    casting_time, recovery_amount_per_second, additional_recovery_amount = bandage[0], bandage[1], bandage[2]
    last_time = attacks[-1][0]
    current_health, attack_index, bandage_count = health, 0, 0
    
    for time in range(1, last_time+1):
        if time < attacks[attack_index][0]:
            bandage_count += 1
            if bandage_count == casting_time:
                current_health += additional_recovery_amount
                bandage_count = 0
            current_health += recovery_amount_per_second
        elif time == attacks[attack_index][0]:
            current_health -= attacks[attack_index][1]
            bandage_count = 0
            if current_health <= 0:
                return -1
            attack_index += 1
        
        if current_health > health:
            current_health = health
            
    return current_health    